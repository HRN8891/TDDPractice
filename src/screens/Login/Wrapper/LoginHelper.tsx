export const performSignIn = () => {
  console.log('Sign In Perform');
};
