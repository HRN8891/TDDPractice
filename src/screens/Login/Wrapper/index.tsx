import React from 'react';
import {
  Keyboard,
  SafeAreaView,
  StatusBar,
  TouchableWithoutFeedback,
  View,
  ViewStyle,
} from 'react-native';
import Style from './style';

interface WrapperProps {
  wrapperStyle?: ViewStyle;
  subContainerStyle?: ViewStyle;
  children: React.ReactNode;
  testId: string;
}

const Wrapper = ({children, wrapperStyle, subContainerStyle, testId}: WrapperProps) => {
  const onContainerClick = () => {
    Keyboard.dismiss();
  };

  return (
    <TouchableWithoutFeedback
      testID={testId}
      style={[Style.container, wrapperStyle]}
      onPress={onContainerClick}>
      <View style={Style.innerView}>
        <StatusBar barStyle="dark-content" />
        <SafeAreaView style={[Style.container, subContainerStyle]}>{children}</SafeAreaView>
      </View>
    </TouchableWithoutFeedback>
  );
};

export default Wrapper;
