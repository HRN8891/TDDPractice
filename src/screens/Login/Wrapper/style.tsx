import {StyleSheet} from 'react-native';
import {COLOR} from '../../../utilities/ColorsType';

const {white} = COLOR.light;
const styles = StyleSheet.create({
  container: {flex: 1},
  subContainer: {
    flex: 1,
  },
  innerView: {
    flex: 1,
    backgroundColor: white,
  },
});

export default styles;
