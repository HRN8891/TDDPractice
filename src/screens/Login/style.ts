import {StyleSheet} from 'react-native';
import {scale} from 'react-native-size-matters';
import {COLOR} from '../../utilities/ColorsType';

const {black1, green, lightGreen2} = COLOR.light;
const styles = StyleSheet.create({
  title: {
    fontSize: scale(20),
    color: black1,
  },
  loginButton: {
    marginTop: scale(25),
  },
  requestCodeHeader: {
    backgroundColor: lightGreen2,
  },
  container: {
    flex: 1,
    marginHorizontal: scale(20),
  },
  emailInput: {
    marginTop: scale(30),
  },
  signupTextView: {
    marginTop: scale(20),
    flexDirection: 'row',
  },
  forgot: {
    color: green,
    fontSize: scale(12),
    marginVertical: scale(25),
  },
  signupText: {
    color: black1,
  },
  signup: {
    color: green,
  },
});
export default styles;
