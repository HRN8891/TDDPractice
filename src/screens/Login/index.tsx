import React, {useCallback, useEffect, useState} from 'react';
import {View} from 'react-native';
import {useDispatch} from 'react-redux';
import CButton from '../../components/CButton';
import CLabel from '../../components/CLabel';
import CTextBox from '../../components/CTextBox';
import Style from './style';
import Wrapper from './Wrapper';

interface Focus {
  [name: string]: boolean;
}

const Login = () => {
  const [userData, setUserData] = useState({
    email: '',
    password: '',
  });
  const [fieldsFocus, setFieldsFocus] = useState<Focus>({
    firstNameFocus: false,
    lastNameFocus: false,
    emailFocus: false,
    passwordFocus: false,
  });
  const [emailError, setEmailError] = useState(false);
  const [passwordError, setPasswordError] = useState(false);
  const dispatch = useDispatch();

  useEffect(() => {}, []);

  const validateForm = useCallback(() => {
    const {email, password} = userData;
    let hasError = false;
    if (!email.trim()) {
      hasError = true;
      setEmailError(true);
    }
    if (!password.trim() || password.length < 8) {
      hasError = true;
      setPasswordError(true);
    }
    return hasError;
  }, [userData]);

  const onSigninClick = useCallback(async () => {}, [dispatch, userData, validateForm]);

  const onChangeText = (value: string, key: string) => {
    setUserData({...userData, [key]: value});
  };

  const clearErrorState = () => {
    setEmailError(false);
    setPasswordError(false);
  };
  const onInputFocus = (focusedKey: string) => {
    clearErrorState();
    setFieldsFocus({
      ...fieldsFocus,
      [focusedKey]: !fieldsFocus[focusedKey],
    });
  };
  const onSignupClick = () => {};

  const onForgotClick = () => {};
  return (
    <Wrapper subContainerStyle={Style.container}>
      <CLabel style={Style.title}>Login</CLabel>
      <CTextBox
        testID="Email"
        inputTitle="Email"
        containerStyle={Style.emailInput}
        value={userData.email}
        onChangeText={value => onChangeText(value, 'email')}
        keyboardType="email-address"
        isFocus={fieldsFocus.emailFocus}
        onFocus={() => onInputFocus('emailFocus')}
        onBlur={() => onInputFocus('emailFocus')}
        error={emailError}
        autoFocus={true}
      />
      <CTextBox
        testID="PasswordTextField"
        inputTitle="Password"
        value={userData.password}
        onChangeText={value => onChangeText(value, 'password')}
        hidePassword={true}
        isFocus={fieldsFocus.passwordFocus}
        onFocus={() => onInputFocus('passwordFocus')}
        onBlur={() => onInputFocus('passwordFocus')}
        error={passwordError}
      />

      {/* {(emailError || passwordError) && <WarningView message={emailError} error />} */}
      <CLabel testID="ForgotButton" onPress={onForgotClick} style={Style.forgot}>
        Forgot password
      </CLabel>
      <CButton testID="LoginButton" onPress={onSigninClick} text="Login" />
      <View style={Style.signupTextView}>
        <CLabel style={Style.signupText}> Not a Member yet? </CLabel>
        <CLabel testID="SignupButton" onPress={onSignupClick} style={Style.signup}>
          SignUp
        </CLabel>
      </View>
    </Wrapper>
  );
};

export default Login;
