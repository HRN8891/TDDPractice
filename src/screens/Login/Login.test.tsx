import React from 'react';
import Login from './index';
import {fireEvent, render} from '@testing-library/react-native';
import {performSignIn} from './Wrapper/LoginHelper';

jest.mock('./Wrapper/LoginHelper', () => ({
  performSignIn: jest.fn(),
}));

// jest.mock('../../src/components/ErrorBoundary', () => jest.fn().mockReturnValue(null));
jest.mock('react-redux', () => {
  return {
    ...jest.requireActual<object>('react-redux'),
    Provider: jest.fn(),
    useSelector: jest.fn(fn => fn()),
    useDispatch: () => jest.fn().mockImplementation(() => Promise.resolve()),
  };
});

// jest.mock('../../src/navigation/root', () => 'navigationRef');

describe('Login screen', () => {
  it('Login screen', () => {
    const tree = render(<Login />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Should render correctly', () => {
    const wrapper = render(<Login />);
    wrapper.getByTestId('Login Component');
  });

  test('Render Password', () => {
    const wrapper = render(<Login />);
    wrapper.getByTestId('Email');
  });

  test('Render Confirm Password', () => {
    const wrapper = render(<Login />);
    wrapper.getByTestId('PasswordTextField');
  });

  test('Render ForgotButton', () => {
    const wrapper = render(<Login />);
    wrapper.getByTestId('ForgotButton');
  });

  test('Render LoginButton', () => {
    const wrapper = render(<Login />);
    wrapper.getByTestId('LoginButton');
  });

  test('Render SignIn Click', () => {
    const wrapper = render(<Login />);
    fireEvent.changeText(wrapper.getByTestId('Email'), 'admin@mail.com');
    fireEvent.changeText(wrapper.getByTestId('PasswordTextField'), 'Admin123');
    fireEvent.press(wrapper.getByTestId('LoginButton'));
    expect(performSignIn).toHaveBeenCalled();
  });
});
