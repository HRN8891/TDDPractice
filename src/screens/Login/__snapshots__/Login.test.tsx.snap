// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login screen Login screen 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#ffffff",
      "flex": 1,
    }
  }
  testID="Login Component"
>
  <View
    style={
      Array [
        Object {
          "flex": 1,
        },
        Object {
          "flex": 1,
          "marginHorizontal": 42.857142857142854,
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "color": "#000000",
          "fontSize": 42.857142857142854,
        }
      }
    >
      Login
    </Text>
    <View
      style={
        Array [
          Object {
            "marginTop": 25.714285714285715,
          },
          Object {
            "marginTop": 64.28571428571428,
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#4E5451",
            "fontSize": 27.857142857142858,
            "marginBottom": 8.571428571428571,
          }
        }
      >
        Email
      </Text>
      <TextInput
        autoCapitalize="none"
        autoFocus={true}
        caretHidden={false}
        keyboardType="email-address"
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        selectionColor="#189E66"
        style={
          Array [
            Object {
              "backgroundColor": "#F1F2F1",
              "borderRadius": 21.428571428571427,
              "fontSize": 32.14285714285714,
              "height": 96.42857142857143,
              "paddingLeft": 10,
              "paddingVertical": 8.571428571428571,
            },
            undefined,
            false,
            undefined,
            false,
          ]
        }
        testID="Email"
        value=""
      />
    </View>
    <View
      style={
        Array [
          Object {
            "marginTop": 25.714285714285715,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#4E5451",
            "fontSize": 27.857142857142858,
            "marginBottom": 8.571428571428571,
          }
        }
      >
        Password
      </Text>
      <TextInput
        autoCapitalize="none"
        caretHidden={false}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        secureTextEntry={true}
        selectionColor="#189E66"
        style={
          Array [
            Object {
              "backgroundColor": "#F1F2F1",
              "borderRadius": 21.428571428571427,
              "fontSize": 32.14285714285714,
              "height": 96.42857142857143,
              "paddingLeft": 10,
              "paddingVertical": 8.571428571428571,
            },
            undefined,
            false,
            undefined,
            false,
          ]
        }
        testID="PasswordTextField"
        value=""
      />
    </View>
    <Text
      accessibilityLabel="ForgotButton"
      onPress={[Function]}
      style={
        Object {
          "color": "#189E66",
          "fontSize": 25.714285714285715,
          "marginVertical": 53.57142857142857,
        }
      }
      testID="ForgotButton"
    >
      Forgot password
    </Text>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "width": "100%",
          },
          undefined,
        ]
      }
    >
      <View
        accessibilityLabel="LoginButton"
        accessibilityRole="button"
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#189E66",
            "borderRadius": 21.428571428571427,
            "justifyContent": "center",
            "opacity": 1,
            "paddingVertical": 27.857142857142858,
            "width": "100%",
          }
        }
        testID="LoginButton"
      >
        <Text
          style={
            Array [
              Object {
                "color": "#ffffff",
                "fontSize": 34.285714285714285,
                "textAlign": "center",
              },
              undefined,
            ]
          }
        >
          Login
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "marginTop": 42.857142857142854,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
          }
        }
      >
         Not a Member yet? 
      </Text>
      <Text
        accessibilityLabel="SignupButton"
        onPress={[Function]}
        style={
          Object {
            "color": "#189E66",
          }
        }
        testID="SignupButton"
      >
        SignUp
      </Text>
    </View>
  </View>
</View>
`;
